classes/core/PKPAppKey.php

```diff
@@ -0,0 +1,185 @@
+     */
+    public static function hasKeyVariable(): bool
+    {
+        return Config::hasVar('general', 'app_key');
+    }
+
+    /**
+     * Get the app key defined in config file
+     */
+    public static function getKey(): string
+    {
+        return Config::getVar('general', 'app_key', '');
+    }
+
+    /**
+     * Validate a given cipher
+     */
+    public static function validateCipher(string $cipher): string
+    {
+        $cipher = strtolower($cipher);
+
+        if (!in_array($cipher, array_keys(static::getSupportedCiphers()))) {
+            $ciphers = implode(', ', array_keys(static::getSupportedCiphers()));
+            
+            throw new Exception(
+                sprintf(
+                    'Invalid cipher %s provided, must be among [%s]',
+                    $cipher,
+                    $ciphers
+                )
+            );
+        }
+
+        return $cipher;
+    }
+
+    /**
+     * Validate given or config defined app
+     */
+    public static function validate(string $key = null, string $cipher = null): bool
+    {
+        $config = app('config')->get('app');
+
+        return Encrypter::supported(
+            static::parseKey($key ?? $config['key']),
+            static::validateCipher($cipher ?? $config['cipher'])
+        );
+    }
+
+    /**
+     * Generate a new app key
+     */
+    public static function generate(string $cipher = null): string
+    {
+        $config = app('config')->get('app');
+
+        return 'base64:'.base64_encode(
+            Encrypter::generateKey(static::validateCipher($cipher ?? $config['cipher']))
+        );
+    }
+
+    /**
+     * Write the given app key in the config file
+     * 
+     * @throws \Exception
+     */
+    public static function writeToConfig(string $key): bool
+    {
+        if (!static::validate($key)) {
+            $ciphers = implode(', ', array_keys(static::getSupportedCiphers()));
+            
+            // Error invalid app key
+            throw new Exception(
+                "Unsupported cipher or incorrect key length. Supported ciphers are: {$ciphers}."
+            );
+        }
+
+        $configParser = new ConfigParser;
+        $configParams = [
+            'general' => [
+                'app_key' => $key,
+            ],
+        ];
+
+        if (!static::hasKeyVariable()) {
+            // Error if the config key `app_key` not defined under `general` section
+            throw new Exception('Config variable named `app_key` not defined in the `general` section');
+        }
+
+        if (!$configParser->updateConfig(Config::getConfigFileName(), $configParams)) {
+            // Error reading config file
+            throw new Exception('Unable to read the config file');
+        }
+
+        if (!$configParser->writeConfig(Config::getConfigFileName())) {
+            // Error writing config file
+            throw new Exception('Unable to write the app key in the config file');
```

Comment:
Since hosts might for legitimate reasons make the config file read-only to the web server, it would be nice to provide a way for admins to manually add the new line of configuration rather than having this be a fatal error. For example, if we successfully generate a key but it can't be written to the configuration file, the upgrade script should instruct the admin what to add to the config file (including the generated key).

That message should also be localized. (It's tough to find a balance between throwing a hard-coded exception with English text vs. localizing the text, but my rule of thumb is: if it's a message that a user might encounter in the normal operation of the system, then it should be localized; if it's a system failure or coding problem, then it's OK to leave it as a hard-coded exception or error. Because admins who are running the system properly might encounter this message, it should be localized.)